sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDbInstance.scala

```diff
@@ -0,0 +1,434 @@
+    verify(readOnly == false, "Cannot open OptimisticTransactionDbInstance in Readonly mode")
+    try {
+      setOptions(conf)
+      options.setCreateIfMissing(true)
+      otdb = OptimisticTransactionDB.open(options, path)
+      dbPath = path
+    } catch {
+      case e: Throwable =>
+        throw new IllegalStateException(
+          s"Error while creating OptimisticTransactionDb instance" +
+            s" ${e.getMessage}",
+          e)
+    }
+  }
+
+  def startTransactions(): Unit = {
+    verify(isOpen(), "Open OptimisticTransactionDbInstance before performing any operation")
+    Option(txn) match {
+      case None =>
+        val optimisticTransactionOptions = new OptimisticTransactionOptions()
+        txn = otdb.beginTransaction(writeOptions, optimisticTransactionOptions)
+        txn.setSavePoint()
+      case Some(x) =>
+        throw new IllegalStateException(s"Already started a transaction")
+    }
+  }
+
+  override def put(key: UnsafeRow, value: UnsafeRow): Unit = {
+    verify(txn != null, "Start Transaction before inserting any key")
+    txn.put(key.getBytes, value.getBytes)
+  }
+
+  override def remove(key: UnsafeRow): Unit = {
+    verify(txn != null, "Start Transaction before deleting any key")
+    txn.delete(key.getBytes)
+  }
+
+  override def get(key: UnsafeRow): UnsafeRow = {
+    verify(txn != null, "Start Transaction before fetching any key-value")
+    Option(txn.get(readOptions, key.getBytes)) match {
+      case Some(valueInBytes) =>
+        val value = new UnsafeRow(valueSchema.fields.length)
+        value.pointTo(valueInBytes, valueInBytes.length)
+        value
+      case None =>
+        null
+    }
+  }
+
+  override def commit(backupPath: Option[String] = None): Unit = {
+    verify(txn != null, "Start Transaction before fetching any key-value")
+    // printTrxStats
+    try {
+      val file = new File(dbPath, identifier.toUpperCase(Locale.ROOT))
+      file.createNewFile()
+      txn.commit()
+      txn.close()
+      txn = null
+      backupPath.foreach(f => createCheckpoint(otdb.asInstanceOf[RocksDB], f))
+    } catch {
+      case e: Exception =>
+        log.error(s"Unable to commit the transactions. Error message = ${e.getMessage}")
+        throw e
+    }
+  }
+
+  def printTrxStats(): Unit = {
+    verify(txn != null, "No open Transaction")
+    logInfo(s"""
+         | deletes = ${txn.getNumDeletes}
+         | numKeys = ${txn.getNumKeys}
+         | puts =  ${txn.getNumPuts}
+         | time =  ${txn.getElapsedTime}
+       """.stripMargin)
+  }
+
+  override def abort(): Unit = {
+    verify(txn != null, "No Transaction to abort")
+    txn.rollbackToSavePoint()
+    txn.close()
+    txn = null
+  }
+
+  override def close(): Unit = {
+    verify(isOpen(), "No DB to close")
+    readOptions.close()
+    writeOptions.close()
+    logDebug("Closing the transaction db")
+    otdb.close()
+    otdb = null
```

Comment:
Shouldn't be a possible transaction closed?

Reply:
Semantic is that transaction will be committed/aborted before close is invoked. So I did not worry about `txn` here.

Reply:
I'm fine with this, but then we can add some `require` here (just in case somebody comes later with a change and doesn't know this decision).